cpp/src/arrow/dataset/dataset.h

```diff
@@ -131,23 +134,23 @@ class ARROW_DS_EXPORT Source {
 /// \brief A Source which yields fragments wrapping a stream of record batches.
 ///
 /// The record batches must match the schema provided to the source at construction.
-class ARROW_DS_EXPORT InMemorySource : public Source {
+class ARROW_DS_EXPORT InMemoryDataset : public Dataset {
  public:
   class RecordBatchGenerator {
    public:
     virtual ~RecordBatchGenerator() = default;
     virtual RecordBatchIterator Get() const = 0;
   };
 
-  InMemorySource(std::shared_ptr<Schema> schema,
-                 std::unique_ptr<RecordBatchGenerator> get_batches)
-      : Source(std::move(schema)), get_batches_(std::move(get_batches)) {}
+  InMemoryDataset(std::shared_ptr<Schema> schema,
```

Comment:
Please create a follow-up jira to implement InMemoryDatasetFactory.

Reply:
https://issues.apache.org/jira/browse/ARROW-7945